File: apps/server2/src/api/utils/api_error_handler.py
from fastapi import Request, status
from fastapi.exceptions import RequestValidationError
from fastapi.responses import JSONResponse

from api.shared.api_error_response import APIErrorResponse
from server_error import Detail, ServerError


class APIErrorHandler:
    def handle_request_validation_error(
        self, request: Request, error: RequestValidationError
    ) -> JSONResponse:
        return JSONResponse(
            content=APIErrorResponse(
                message="Validation failed",
                details=Detail(context=error.body, cause=error.errors()),
                is_operational=True,
            ).model_dump(),
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
        )

    def handle_server_error(self, request: Request, error: ServerError) -> JSONResponse:
        return JSONResponse(
            content=APIErrorResponse(
                message=error.message,
                details=error.detail,
                is_operational=error.is_operational,
            ).model_dump(),
            status_code=error.status_code,
        )

File: apps/server1/src/api/utils/api-error-handler.ts
import { NextFunction, Request, Response } from 'express'
import httpStatus from 'http-status'
import { ValidateError } from 'tsoa'

import { Detail, ServerError } from '../../server-error'
import { APIErrorResponse } from '../shared'

class APIErrorHandler {
  public handleError(
    error: unknown,
    req: Request,
    res: Response,
    next: NextFunction,
  ): Response | void {
    if (error instanceof ValidateError) {
      const response: APIErrorResponse = {
        message: 'Validation failed',
        detail: { context: error?.fields, cause: error },
        isOperational: true,
      }
      return res.status(httpStatus.UNPROCESSABLE_ENTITY).json(response)
    }

    if (error instanceof ServerError) {
      const response: APIErrorResponse = {
        message: error.message,
        detail: error.newDetail,
        isOperational: error.isOperational,
      }
      return res.status(error.statusCode).json(response)
    }

    if (error instanceof Error) {
      const response: APIErrorResponse = {
        message: error.message,
        detail: { context: undefined, cause: error } as Detail,
        isOperational: false,
      }
      res.status(httpStatus.INTERNAL_SERVER_ERROR).json(response)
      return
    }

    next()
  }
}

export { APIErrorHandler }
